feat(agent,sysdig-deploy): Implement global plan settings

aroberts87 opened the pull request:
## What this PR does / why we need it:
Introduce a new 'Sysdig Aware' subchart enable/disable feature that leverages global.sysdig.monitor and global.sysdig.secure to toggle:
1) Which subcharts of sysdig-deploy get rendered based on Sysdig deployment type
2) Which Agent settings are automatically set based on the type of deployment configured by the user.

This change also adds logic to enable secure_light mode in the Agent when specified.

The implementation here allows for the legacy <chart>.enabled values to override the new plan settings in an attempt at being backwards compatible.

## Checklist

[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]

- [x] Title of the PR starts with type and scope, (e.g. feat(agent,node-analyzer,sysdig-deploy):)
- [x] Chart Version bumped for the respective charts
- [ ] Variables are documented in the README.md (or README.tpl in some charts)
- [x] Check GithubAction checks (like lint) to avoid merge-check stoppers
- [x] All test files are added in the tests folder of their respective chart and have a "_test" suffix

Check Contribution guidelines in README.md for more insight.

aroberts87 committed:
Implement global plan settings

Introduce a new 'Sysdig Aware' subchart enable/disable feature that leverages
global.sysdig.monitor and global.sysdig.secure to toggle:
1) Which subcharts of sysdig-deploy get rendered based on Sysdig deployment type
2) Which Agent settings are automatically set based on the type of deployment
   configured by the user.
2a) Enabled support for global.sysdig.secure when setting
    agent.sysdig.settings.feature.mode to secure_light

The implementation here allows for the legacy <chart>.enabled values to
override the new plan settings in an attempt at being backwards compatible.

balinaveen reviewed (commented):
What about changes to Readme? I don't see the new global plan settings documented. 
It's not clear how secure_light should be enabled. It there a global plan setting that maps to this mode or is this an agent sub chart setting only?

aroberts87 commented:
> What about changes to Readme? I don't see the new global plan settings documented. It's not clear how secure_light should be enabled. It there a global plan setting that maps to this mode or is this an agent sub chart setting only?

You beat me too it, @balinaveen. I didn't realize I was missing the README changes until I opened the PR. I'm making those changes now, but wanted to keep the PR open so people could take a look at what's there.

aroberts87 committed:
Update READMEs for global plan types and secure_light

lilx1ao commented on charts/sysdig-deploy/README.md:
````diff
@@ -203,6 +205,42 @@ The following table lists the configurable parameters of the sysdig-deploy chart
+```bash
+$ helm install --namespace sysdig --set agent.sysdig.settings.features.mode=secure_light sysdig sysdig/sysdig-deploy
+```
+
````
nice read 👍

lilx1ao commented on charts/agent/README.md:
````diff
@@ -277,6 +277,19 @@ $ helm install --namespace sysdig-agent sysdig-agent -f values.yaml sysdig/agent
+
+To deploy the Sysdig Agent in Secure Light mode, run:
+```bash
+$ helm install --namespace sysdig-agent sysdig-agent --set sysdig.settings.features.mode=secure_light sysdig/agent
````
why did you pick underscore over hyphen tho?

aroberts87 replied:
Not my choice. 🙂 Anything under `sysdig.settings` is directly written into the ConfigMap and the agent itself uses
```
features:
  mode: secure_light
```
https://docs.sysdig.com/en/docs/installation/sysdig-agent/agent-configuration/configure-agent-modes/#secure-light

lilx1ao replied:
I see!

lilx1ao commented on charts/agent/templates/_helpers.tpl:
```diff
@@ -319,6 +319,58 @@ Use global sysdig tags for agent
+    {{- if and (and (hasKey .Values.global.sysdig "secure")
+                    (not .Values.global.sysdig.secure))
+                    .Values.secureFeatProvided }}
+        {{ fail "Cannot unset global.sysdig.secure when agent.sysdig.settings.feature.mode is secure or secure_light" }}
```
I would add a suggestion "set global.sysdig.secure to be enabled" etc

aroberts87 committed:
Adjust README files

- Removed mentions of the Agent secure_light mode
- Added mention in sysdig-deploy README that enabling both monitor and secure
  modes equates to a platform installation

aroberts87 committed:
Update error message for improper plan settings

lilx1ao commented on charts/sysdig-deploy/Chart.yaml:
```diff
@@ -21,28 +21,28 @@ dependencies:
     version: ~0.7.23
     alias: admissionController
-    condition: admissionController.enabled
+    condition: admissionController.enabled,global.sysdig.plan.secure
```
what's .plan?

aroberts87 replied:
Good catch. 😬

lilx1ao commented on charts/agent/templates/_helpers.tpl:
```diff
@@ -319,6 +319,58 @@ Use global sysdig tags for agent
+            "jmx"
+            "prometheus"
+            "statsd") }}
+            {{- $_ := set $monitorBlock $monitorFeature (dict "enabled" false) }}
```
I can guess what this block is trying to accomplish but my brain isn't allowing me to look too close today. as long as it's tested we cool

aroberts87 replied:
No worries, it's not trivial. It's assembling a dictionary that then gets rendered as YAML.
```yaml
app_checks_enabled: false
jmx:
  enabled: false
prometheus:
  enabled: false
statsd:
 enabled: false
```
I tested it regularly during development by manually inspecting the generated manifests, and there is [this check](https://github.com/sysdiglabs/charts/pull/911/files/5472a66391c3c1f31e15f193a0a0176e1bc7ffd3#diff-e33b13b6f8de45bf1e25d748554be6312f747666ae850b66c88d5a71c0638fc8R90-R93) for `app_checks_enabled: false` when secure only mode, but I can extend it to capture the whole block.

aroberts87 committed:
Fix chart dependency conditions to match revised key

balinaveen commented on charts/sysdig-deploy/README.md:
```diff
@@ -203,6 +205,29 @@ The following table lists the configurable parameters of the sysdig-deploy chart
 | `rapidResponse`                         | Config specific to [Sysdig Rapid Response](#rapid-response)                                                             | `{}`      |
 | `rapidResponse.enabled`                 | Enable Rapid Response component in this chart                                                                           | `""`      |
 
+## Sysdig Plan Settings
```
The onboarding page in the UI shows this under "helm"

```
kubectl create ns sysdig-agent
helm repo add sysdig https://charts.sysdig.com
helm repo update
helm install sysdig-agent --namespace sysdig-agent \
    --set global.sysdig.accessKey=e64a1780-5bbd-4083-8ce1-418c2bb0a356 \
    --set global.sysdig.region=custom \
    --set nodeAnalyzer.secure.vulnerabilityManagement.newEngineOnly=true \
    --set global.kspm.deploy=true \
    --set nodeAnalyzer.nodeAnalyzer.benchmarkRunner.deploy=false \
    --set global.clusterConfig.name=<CLUSTER_NAME> \
    sysdig/sysdig-deploy 
```
How will that change with the introduction of the new global plan settings? For example, are the nodeAnalyzer settings still required?

aroberts87 replied:
With these changes as they are the instructions could become slightly more complicated as secure customers can now optionally set `global.sysdig.secure=true` in the `helm install ...` command. The reason for this is because the other settings, particularly `nodeAnalyzer.secure.vulerabilityManagement.newEngineOnly=true` and `nodeAnalyzer.nodeAnalyzer.benchmarkRunner.deploy=false` are manipulating how the subcharts work.

One change I can do is include the the KSPM collector as a part of `global.sysdig.secure=true`. There was actually a whole discussion on this topic yesterday (disabling bencmark v2 and enabling kspm-collector when `global.sysdig.secure` is set). I was planning on doing that change in a follow up PR, but it wouldn't be too hard to put into this one. @balinaveen , what are your thoughts?

aroberts87 replied:
>One change I can do is include the the KSPM collector as a part of global.sysdig.secure=true. There was actually a whole discussion on this topic yesterday (disabling bencmark v2 and enabling kspm-collector when global.sysdig.secure is set). I was planning on doing that change in a follow up PR, but it wouldn't be too hard to put into this one. @balinaveen , what are your thoughts?

On second thought, I would not like to do this as a part of this PR. I would like to engage the KSPM folks for this change so that they are aware.

balinaveen replied:
Yeah, let's do the ksmp changes later. If the onboarding UI as it exists today will still work, then that's OK. I was hoping things would be simpler, but that can wait. Let's postpone the update to the UI until all changes needed to simplify things are in the helm chart.

claffin replied:
We should implement this with KSPM collector and KSPM analyzer as a part of global.sysdig.secure=true.

KSPM is now GA and the benchmark runner is now deprecated. We shouldn't give these plans to customers without this otherwise they'll be missing features, especially new customers during a PoV/onboarding.

aroberts87 commented:
Based on some discussions about this PR we have opted to close this single PR and instead do the following:
1) Close this PR and break out the agent chart specific improvements and get those in a separate PR.
2) Wait for the benchmark runner --> kspm migration to happen
3) Open a new PR for these plan setting improvements after the aforementioned migration has occurred.
